### ISSUE TITLE
400 Bad Request whenever binary=true

### ISSUE THREAD

**Issue by sk89q** [miguelgrinberg/Flask-SocketIO#234]
I'm getting a 400 Bad Request whenever `binary` is `True` on `SocketIO()`.

server.py:

```python
#!/usr/bin/env python2.7

import logging
from flask import Flask
from flask_socketio import SocketIO, emit

app = Flask(__name__)
app.debug = False
app.config['SECRET_KEY'] = 'secret!'
socketio = SocketIO(app, binary=True)

@socketio.on('message')
def handle_message(message):
    emit("message", message)

if __name__ == '__main__':
    logging.basicConfig(format='%(levelname)s [%(name)s] %(message)s', level=logging.DEBUG)
    socketio.run(app, host="0.0.0.0", port=5000)
```

client.html:

```html
<!doctype html>
<html>
<body>
<script src="https://cdn.socket.io/socket.io-1.4.5.js"></script>
<script>
  var socket = io.connect('http://' + document.domain + ':' + location.port);
  socket.on("connect", function() {
    socket.emit("message", {});
  });
</script>
</body>
</html>
```

Google Chrome console output:

```
GET http://localhost:5000/socket.io/?EIO=3&transport=polling&t=LCBOV0o&sid=5c3e44746b11444eb7c8c5712d01a717 net::ERR_CONNECTION_REFUSED
WebSocket connection to 'ws://localhost:5000/socket.io/?EIO=3&transport=websocket&sid=5c3e44746b11444eb7c8c5712d01a717' failed: WebSocket is closed before the connection is established.
```

Python logging output (if async_mode=gevent):

```
DEBUG [geventwebsocket.handler] Initializing WebSocket
DEBUG [geventwebsocket.handler] Validating WebSocket request
DEBUG [geventwebsocket.handler] Can only upgrade connection if using GET method.
```

Tested on:

* Python 2.7
* Windows 8.1
* async_mode could be either eventlet, gevent, or threaded and it still occurs

```
eventlet==0.18.4
Flask==0.10.1
Flask-SocketIO==2.1
gevent==1.0.2
gevent-websocket==0.9.5
greenlet==0.4.9
itsdangerous==0.24
Jinja2==2.8
MarkupSafe==0.23
python-engineio==0.8.8
python-socketio==1.0
six==1.10.0
Werkzeug==0.11.4
wheel==0.29.0```

**Comment by miguelgrinberg** [comment]
I'll look into it. The `binary=True` is really a hack to work around the limitations in Python 2 with dealing with text vs. binary. When you work in that mode, all strings that are not unicode will be treated as binary. I suspect that's not a rule that is respected end to end.

In any case, if you plan to work with binary content, I strong advise you to use Python 3 if you can. Even with this bug fixed, you'll have to always ensure your text content is always unicode.

**Comment by sk89q** [comment]
OK so I tested it on Python 3.5.1 and I have the exact same problem.

In summary, I can reproduce this problem on:

* Python 2.7.11 on Windows 8.1
* Python 3.5.1 on Windows 8.1
* Python 3.4.3 on Ubuntu 14.04

**Comment by miguelgrinberg** [comment]
Can you share your application so that I can try it here?

**Comment by sk89q** [comment]
The application is pretty large so I've narrowed down the issue to just sending binary data.

This results in a 400 Bad Request for me:

```html
<!doctype html>
<html>
<body>
<script src="https://cdn.socket.io/socket.io-1.4.5.js"></script>
<script>
  var socket = io.connect('http://' + document.domain + ':' + location.port);
  socket.on("connect", function() {
    socket.emit("request", "Give it to me");
  });
  socket.on("response", function(data) {
    console.log("got " + data)
  });
</script>
</body>
</html>
```

```python
import logging

from flask import Flask
from flask_socketio import SocketIO, emit

app = Flask(__name__)
app.debug = False
app.config['SECRET_KEY'] = 'secret!'
socketio = SocketIO(app, binary=True)

@socketio.on('request')
def handle_message(message):
    emit("response", b'\x05\x06')

if __name__ == '__main__':
    logging.basicConfig(format='%(levelname)s [%(name)s] %(message)s', level=logging.DEBUG)
    socketio.run(app, host="0.0.0.0", port=8044)
```

Sending text works for me:

```python
@socketio.on('request')
def handle_message(message):
    emit("response", "Hello")
```

Am I doing something wrong?

In the actual application, if binary data isn't set near initial connection time, it still fails but rather than the 400 Bad Request, socket.io gives me an error callback with the message "server error".

**Comment by miguelgrinberg** [comment]
Thanks for providing an example.

It looks like there is a problem with binary packets when using long-polling. Your example sends the binary data right away after a connection, so this happens before the upgrade to websocket is complete. I changed your template a little bit to see if I can avoid the problem if I send the data after websocket is active, and that seems to work. The change I did is to delay the sending of the initial packet by 5 seconds:

    socket.on("connect", function() {
        window.setTimeout(function() { socket.emit("request", "Give it to me"); }, 5000);
    });

So I need to figure out what the binary packets fail when doing long-polling. I'll update the issue once I have more details.

**Comment by miguelgrinberg** [comment]
@sk89q I believe the current master branch of python-engineio addresses this problem. If you are using Python 3.x, then you can initialize your `SocketIO` normally and then pass a bytes object to have it sent to the client as a binary packet:

    socketio = SocketIO(app)
    # ...
    emit("response", b'\x05\x06')

In Python 2.x, you have to use the explicit binary mode, and then work with unicode strings for everything that needs to be sent as text:

    socketio = SocketIO(app, binary=True)
    # ...
    emit(u"response", b'\x05\x06')

Could you please install python-engineio from the git repository and confirm that we can get this to work please?

**Comment by miguelgrinberg** [comment]
Fixed in python-engineio package, version 0.9.0. Please let me know if you continue to see problems after upgrade to that release.
